State diff

winsvega opened the pull request:
add --statediff option when filling or running tests

gumb0 commented on libethereum/State.cpp:
```diff
@@ -541,6 +541,7 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
 	u256 const startGasUsed = _envInfo.gasUsed();
 	executeTransaction(e, _t, onOp);
 
+	res.changeLog = m_changeLog;
```
I don'like that this way we will have this in `ExecutionResult` (paying with additional memory consumption) even when we don't need it (we need it only for tests and maybe vmtraces)

@chfast what do you think?

winsvega replied:
could be under cmake flag --VMTRACE

chfast replied:
I don't like it either. This must go to `dev::test` namespace unless proven it is not possible to be done this way.

And big no to build flags. Let's find a way to do it nicely without affecting not-test related code.

gumb0 replied:
I had in mind adding another callback function for collecting state diffs as `State::execute` parameter.

chfast replied:
Is the "state diff" the same as "state journal" or "state changelog"? If yes, let's stick to a single name. We can add a function to export the changelog next to the functions like `savepoint()` or `revert()`.

gumb0 replied:
I think it's essentially the same, let's call it change log.

We need it after transaction execution (`executeTransaction` call on line 542), but before `commit` which clears changelog.

The getter function would work if we save the copy before commit, or don't clear it.

winsvega replied:
yes, getter function would be nice. 
state.executeTransaction()
state.getChangelog()

gumb0 commented on libdevcore/CommonData.h:
```diff
@@ -172,9 +179,9 @@ inline std::string toHex(u256 val, HexPrefix prefix = HexPrefix::DontAdd)
 }
 
-inline std::string toCompactHex(u256 val, HexPrefix prefix = HexPrefix::DontAdd, unsigned _min = 0)
+inline std::string toCompactHex(u256 val, HexPrefix prefix = HexPrefix::DontAdd, PrintZero _min = PrintZero::DontPrint)
```
So many places where we pass `PrintZero::Print` here, is it ever used really with `PrintZero::DontPrint` ?

chfast replied:
This looks bad. Can you describe what the use cases for this function are? If tests needs something different maybe it will be better for `dev::test` to have a separated function for this case.

winsvega replied:
there is a person who constantly asking me to change hex format in our tests. I've changed it here to PrintZero::DontPrint because that is what the meaning of that "1" was in the code. 
but the changes are not final.

gumb0 commented on test/tools/libtesteth/ImportTest.cpp:
```diff
@@ -629,6 +654,10 @@ int ImportTest::exportTest(bytes const& _output)
 					obj2["postState"] = fillJsonWithState(m_transactions[i].postState);
 			}
+
+			if (Options::get().statediff)
```
Does this go into generated json files?

winsvega replied:
when using with --filltests yes.

gumb0 replied:
Do we really need to write it into files and incerase their size and complexity? It's only for human readers, right?

Maybe just leave it like a log option for `testeth` run ?

chfast replied:
Answer for this?

gumb0 commented on test/tools/libtesteth/TestHelper.cpp:
```diff
@@ -124,19 +125,126 @@ eth::Network stringToNetId(string const& _netname)
+	std::vector<dev::eth::detail::Change> changeLog = _changeLog;
+
+	std::ostringstream log;
+	while(changeLog.size())
```
Why do you need to iterate in this weird way instead of just `for (auto& change : changeLog)` ?

winsvega replied:
I use this because of the changeLog.erase

gumb0 replied:
Why do you need to erase while iterating over it?

gumb0 replied:
If you need group changes by the addresses, you could sort `changeLog` vector by `address` field, then do all the processing in one iteration over the vector

gumb0 commented:
It probably would be nice to have this feature not only for tests, but also for regular `eth` operation, like we have vmtrace for both... Would it be difficult to output state diff to log in `eth` when, say, VMTRACE is enabled ?

chfast commented:
Ok, let's take is slowly.
Can we separate the changes related to `toCompactHex` from the "state diff" one first?

winsvega commented:
refactored the log function. 
I've removed changeLog clean function from commit in state. and moved it to the begining  of execute.
will see how tests work. if there is an issue we could dublicate m_changeLog variable in State just for  getChangeLog() method

gumb0 commented on test/tools/libtesteth/FillJsonFunctions.cpp:
```diff
@@ -0,0 +1,228 @@
+	changeStruct popChange = changeLog.at(0);
+	std::map<string, ostringstream> logInfo;
+	u256 tmpBalance = _stateOrig.balance(popChange.address);
+	for (size_t i = 0; i < changeLog.size(); i++)
```
This loop probably works correctly, but I think expressing it something like this would be clearer:
```
size_t i = 0;
while (i < changeLog.size())
{
	Addres currentAddress = changeLog[i].address;

	string before;
	string after;
	json_spirit::mArray record;
	json_spirit::mArray agregatedBalance;
	json_spirit::mArray agregatedStorage;
	std::map<string, ostringstream> logInfo;

	while (i < changeLog.size() && changeLog[i].address == currentAddress)
	{
		// handle changeLog[i]

		++i;
	}
}
```

chfast replied:
@winsvega ?

pirapira commented:
I think ChangLog.md can talk about the new option.

pirapira commented on test/tools/libtesteth/FillJsonFunctions.cpp:
```diff
@@ -0,0 +1,228 @@
+			o["balance"] = toCompactHex(_state.balance(a.first), HexPrefix::Add, 1);
+		if (mapEmpty || _map.at(a.first).hasNonce())
+			o["nonce"] = toCompactHex(_state.getNonce(a.first), HexPrefix::Add, 1);
+		{
```
Why is a `{` here?  The `if` statement has ended in the last line.

pirapira commented on test/tools/libtesteth/ImportTest.cpp:
```diff
@@ -637,6 +667,8 @@ int ImportTest::exportTest(bytes const& _output)
 
 		m_testObject["post"] = obj;
+
+
```
Spurious lines?

chfast commented on libethereum/StateDetail.h:
```diff
@@ -0,0 +1,75 @@
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file StateDetail.h
```
Why did you move it to another file?

chfast replied:
You have to move it back.

winsvega commented:
the first solution was to put changeLog into execution results. that required separation of StateDetail from State

chfast commented on test/tools/libtesteth/FillJsonFunctions.cpp:
```diff
@@ -0,0 +1,227 @@
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file
```
What is the point of not having these function in TestHelper.cpp?

winsvega commented:
to split the large file with lots of functions into smaller ones

chfast commented:
This just makes it compile twice as long as before.

winsvega commented:
I stick to that guy's opinion:
"f I find that an implementation file is thousands of lines long, that's usually a sign that there's too much there and I need to break it up."
https://stackoverflow.com/questions/531133/should-i-put-many-functions-into-one-file-or-more-or-less-one-function-per-fi

chfast commented on libethereum/State.cpp:
```diff
@@ -224,7 +225,6 @@ void State::commit(CommitBehaviour _commitBehaviour)
 	if (_commitBehaviour == CommitBehaviour::RemoveEmptyAccounts)
 		removeEmptyAccounts();
 	m_touched += dev::eth::commit(m_cache, m_state);
-	m_changeLog.clear();
```
`commit()` have to clear the changelog. See comments below.

chfast commented on test/tools/libtesteth/TestHelper.cpp:
```diff
@@ -477,7 +422,10 @@ void executeTests(const string& _name, const string& _testPathAppendix, const st
 	{
-		cnote << "TEST " << name << ":";
+		if ((Options::get().singleTest && Options::get().singleTestName == name)
+			||	!Options::get().singleTest)
```
WTF is this tab in the middle?

winsvega commented:
staediff also work when filling tests producing output to the test .json for debug

chfast commented on libethereum/State.cpp:
```diff
@@ -223,9 +224,12 @@ void State::commit(CommitBehaviour _commitBehaviour)
 	m_touched += dev::eth::commit(m_cache, m_state);
-	m_changeLog.clear();
+	if (m_changeLog.size() != 0)
+		m_changeLogResult = m_changeLog;
```
This is bad and inefficient (both fixable), but read below.

chfast commented on test/tools/libtesteth/ImportTest.cpp:
```diff
@@ -93,6 +93,7 @@ bytes ImportTest::executeTest()
 				eth::Network network = networks[j];
 				std::pair<eth::State, ImportTest::execOutput> out = executeTransaction(network, m_envInfo, m_statePre, m_transactions[i].transaction);
```
You should try to improve `executeTransaction()` function first.
1. Do you need preState after this function. If not do not copy the State.
2. Can it also return changelog if you need it?
3. Use `std::tie()` to unpack tuple.

chfast commented on libethereum/State.cpp:
```diff
@@ -223,9 +223,10 @@ void State::commit(CommitBehaviour _commitBehaviour)
 	m_touched += dev::eth::commit(m_cache, m_state);
-	m_changeLog.clear();
 	m_cache.clear();
+	m_changeLog.clear();
```
Revert all changes here.

chfast commented on libethcore/ChainOperationParams.h:
```diff
@@ -15,8 +15,6 @@
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
 /** @file ChainOperationsParams.h
- * @author Gav Wood <i@gavwood.com>
```
Do not do this unrelated changes.

chfast commented on libethereum/State.cpp:
```diff
@@ -541,12 +542,18 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
-		commit(removeEmptyAccounts ? State::CommitBehaviour::RemoveEmptyAccounts : State::CommitBehaviour::KeepEmptyAccounts);
+		case Permanence::Reverted:
+			m_cache.clear();
+		break;
```
`break` must be indented.

chfast commented on libethereum/State.cpp:
```diff
@@ -541,12 +542,18 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
+			removeEmptyAccounts = _envInfo.number() >= _sealEngine.chainParams().u256Param("EIP158ForkBlock");
+			commit(removeEmptyAccounts ? State::CommitBehaviour::RemoveEmptyAccounts : State::CommitBehaviour::KeepEmptyAccounts);
+		break;
+		default:
```
Use `Unfinished` instead of `default`. Compilers will be satisfied by having cases for all options.

chfast commented on libethereum/State.h:
```diff
@@ -15,8 +15,6 @@
 	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
 /** @file State.h
- * @author Gav Wood <i@gavwood.com>
```
Unrelated change.

chfast commented on libethereum/State.h:
```diff
@@ -82,7 +80,8 @@ enum class BaseState
 	Reverted,
-	Committed
+	Committed,
+	Unfinished
```
Documentation missing.

chfast commented on libethereum/State.h:
```diff
@@ -142,8 +141,7 @@ struct Change
 
-}
-
+} //detail
```
Ugly formatting. Put 2 spaces before `//`, and space after `//`. Also better comment: `// namespace detail`.

chfast commented on libethereum/State.h:
```diff
@@ -336,7 +336,7 @@ class State
 
 	friend std::ostream& operator<<(std::ostream& _out, State const& _s);
-	std::vector<detail::Change> m_changeLog;
+	std::vector<detail::Change> m_changeLog;	///< Working changeLog
```
1. Working means the one that is not broken. You mean current, or change log of currently executing transaction?
2. "changeLog" is not an English word. You mean "change log"?
3. Missing full stop.

chfast commented on test/tools/libtesteth/ImportTest.cpp:
```diff
@@ -52,13 +53,14 @@ ImportTest::ImportTest(json_spirit::mObject& _o, bool isFiller, testType testTem
 
 bytes ImportTest::executeTest()
 {
+	using results = std::tuple<eth::State, ImportTest::ExecOutput, ImportTest::ChangeLog>;
```
Type names must start with uppercase letter. Should be `Results`.

chfast commented on test/tools/libtesteth/ImportTest.cpp:
```diff
@@ -52,13 +53,14 @@ ImportTest::ImportTest(json_spirit::mObject& _o, bool isFiller, testType testTem
 bytes ImportTest::executeTest()
 {
+	using results = std::tuple<eth::State, ImportTest::ExecOutput, ImportTest::ChangeLog>;
 	if (m_testType == testType::StateTests)
```
Is this case needed any more?

chfast commented on test/tools/libtesteth/ImportTest.cpp:
```diff
@@ -91,8 +93,9 @@ bytes ImportTest::executeTest()
 				eth::Network network = networks[j];
-				std::pair<eth::State, ImportTest::execOutput> out = executeTransaction(network, m_envInfo, m_statePre, m_transactions[i].transaction);
-				m_transactions[i].postState = out.first;
+				results out = executeTransaction(network, m_envInfo, m_statePre, m_transactions[i].transaction);
```
Use `std::tie()` to assign `postState` and `changeLog` and ignore tuple element of index 1.

chfast commented on test/tools/libtesteth/ImportTest.h:
```diff
@@ -54,15 +54,17 @@ class ImportTest
 private:
-	typedef std::pair<eth::ExecutionResult, eth::TransactionReceipt> execOutput;
-	std::pair<eth::State, execOutput> executeTransaction(eth::Network const _sealEngineNetwork, eth::EnvInfo const& _env, eth::State const& _preState, eth::Transaction const& _tr);
+	typedef std::vector<dev::eth::detail::Change> ChangeLog;
```
This must go to State.h. Also use it in State class declaration.

chfast commented on test/tools/libtesteth/ImportTest.h:
```diff
@@ -54,15 +54,17 @@ class ImportTest
-	typedef std::pair<eth::ExecutionResult, eth::TransactionReceipt> execOutput;
-	std::pair<eth::State, execOutput> executeTransaction(eth::Network const _sealEngineNetwork, eth::EnvInfo const& _env, eth::State const& _preState, eth::Transaction const& _tr);
+	typedef std::vector<dev::eth::detail::Change> ChangeLog;
+	typedef std::pair<eth::ExecutionResult, eth::TransactionReceipt> ExecOutput;
```
Use `using` instead of `typedef` to make a type alias.

chfast commented on test/tools/libtesteth/TestHelper.cpp:
```diff
@@ -477,7 +422,9 @@ void executeTests(const string& _name, const string& _testPathAppendix, const st
 	{
-		cnote << "TEST " << name << ":";
+		if ((Options::get().singleTest && Options::get().singleTestName == name) || !Options::get().singleTest)
+			cnote << "TEST " << name << ":";
```
Verbosity 2 outputs test names. Why do you need this?

winsvega replied:
so that when you use --statediff you would see only the name of the test you are running with --singletest option. 
other names make it hard to see the changeLog in console when there are many tests

chfast commented on libethereum/State.cpp:
```diff
@@ -541,12 +541,18 @@ std::pair<ExecutionResult, TransactionReceipt> State::execute(EnvInfo const& _en
-	if (_p == Permanence::Reverted)
-		m_cache.clear();
-	else
+	bool removeEmptyAccounts;
```
This variable must be initialized.

chfast commented on libethereum/State.h:
```diff
@@ -82,7 +82,8 @@ enum class BaseState
 	Reverted,
-	Committed
+	Committed,
+	Unfinished	///< uncommited state for changeLog readings
```
In documenting comments always use full sentences. Must be `Uncommited state for change log readings in tests.".
Seeing the comment, shouldn't we change `Unfinished` to `Uncommited`?

chfast commented on libethereum/State.h:
```diff
@@ -142,8 +143,9 @@ struct Change
 };
 
-}
+typedef std::vector<Change> ChangeLog;
```
1. `using ChangeLog = std::vector<Change>;
2. It must not be in `detail` namespace. Put it in `eth` namespace directly. Now the `detail` namespace is pointless as we now make the change log available outside of State.

chfast commented on test/tools/libtesteth/ImportTest.cpp:
```diff
@@ -55,10 +56,13 @@ bytes ImportTest::executeTest()
+		LogEntries emptyLogs;
+		ExecutionResult emptyRes;
+		ImportTest::ExecOutput execOut = make_pair(emptyRes, TransactionReceipt(h256(), u256(), emptyLogs));
+		std::tie(m_statePost, execOut, std::ignore) =
```
Here this change was not necessary as ExecutionResult does not have default constructor, but it can stay.

winsvega replied:
we will remove this part anyway at some point

chfast commented on libethereum/State.h:
```diff
@@ -82,7 +82,8 @@ enum class BaseState
 	Reverted,
-	Committed
+	Committed,
+	Uncommited	///< Uncommited state for change log readings in tests.
```
Typo, `Uncommitted`.

chfast commented on test/tools/libtesteth/FillJsonFunctions.cpp:
```diff
@@ -0,0 +1,218 @@
+	if (!_changeLog.size())
+		return oState;
+
+	using changeKind = eth::Change::Kind;
```
Type aliases must start from upper letter.

chfast commented on test/tools/libtesteth/FillJsonFunctions.cpp:
```diff
@@ -52,19 +52,19 @@ json_spirit::mObject fillJsonWithTransaction(Transaction const& _txn)
-	using changeStruct = eth::detail::Change;
-	eth::detail::ChangeLog changeLog = _changeLog;
+	using ChangeKind = eth::Change::Kind;
+	using ChangeStruct = eth::Change;
```
This make no sense. You can use `Change` (because `using namespace dev::eth`), why you want to replace it with `ChangeStruct`.

Similar `ChangeKind` instead of `Change::Kind`?

winsvega replied:
because we use mutation approach on this PR

chfast commented on test/tools/libtesteth/FillJsonFunctions.cpp:
```diff
@@ -52,19 +52,19 @@ json_spirit::mObject fillJsonWithTransaction(Transaction const& _txn)
 	//Sort the vector by address field
 	std::sort(changeLog.begin(), changeLog.end(),
-		[ ]( const changeStruct& lhs, const changeStruct& rhs )
+		[ ]( const ChangeStruct& lhs, const ChangeStruct& rhs )
```
Too much spaces here. Should be: `[](const ChangeStruct& lhs, const ChangeStruct& rhs)`.
